File: 백준/Silver/2343. 기타 레슨/기타 레슨.c
#include <stdio.h>
#include <stdlib.h>
int main(void) {
    int n, m, left = 0, right = 0, mid, cnt, total;
    scanf("%d %d", &n, &m);
    int *disc = (int *)malloc(sizeof(int) * n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &disc[i]);
        if (left < disc[i])
            left = disc[i];
        right += disc[i];
    }
    while (left <= right) {
        mid = (left + right) / 2;
        cnt = 1, total = 0;
        for (int i = 0; i < n; i++) {
            if (total + disc[i] <= mid) {
                total += disc[i];
            } else {
                total = disc[i];
                cnt += 1;
            }
        }
        if (m < cnt)
            left = mid + 1;
        else
            right = mid - 1;
    }
    printf("%d", left);
    return 0;
}

File: 백준/Silver/2343. 기타 레슨/기타 레슨.py
import sys
input=lambda:sys.stdin.readline().rstrip()
n,m=map(int,input().split())
lessons=list(map(int,input().split()))
def can_record(lessons,m,size):
    cnt,cur_sum=1,0
    for lesson in lessons:
        if lesson>size:
            return False
        if cur_sum+lesson>size:
            cnt+=1
            cur_sum=0
        cur_sum+=lesson
    return cnt<=m
def min_bluray_size(n,m,lessons):
    left,right=max(lessons),sum(lessons)
    result=right
    while left<=right:
        mid=(left+right)//2
        if can_record(lessons,m,mid):
            result=mid
            right=mid-1
        else:
            left=mid+1
    return result
print(min_bluray_size(n,m,lessons))